libs/vkd3d-shader/spirv.c

```diff
@@ -5105,8 +5113,31 @@ static const struct vkd3d_shader_global_binding *vkd3d_dxbc_compiler_get_global_
     {
         if (binding->flags & VKD3D_SHADER_BINDING_FLAG_COUNTER)
         {
-            ERR("Bindless UAV counters not supported.\n");
-            return NULL;
+            uint32_t counter_struct_id, pointer_struct_id, array_type_id;
+
+            counter_struct_id = vkd3d_spirv_get_type_id(builder, VKD3D_TYPE_UINT, 1);
+            counter_struct_id = vkd3d_spirv_build_op_type_struct(builder, &counter_struct_id, 1);
+
+            vkd3d_spirv_build_op_member_decorate1(builder, counter_struct_id, 0, SpvDecorationOffset, 0);
+            vkd3d_spirv_build_op_decorate(builder, counter_struct_id, SpvDecorationBlock, NULL, 0);
+            vkd3d_spirv_build_op_name(builder, counter_struct_id, "uav_ctr_t");
+
+            type_id = vkd3d_spirv_build_op_type_pointer(builder, SpvStorageClassPhysicalStorageBuffer, counter_struct_id);
+
+            array_type_id = vkd3d_spirv_build_op_type_runtime_array(builder, type_id);
+            vkd3d_spirv_build_op_decorate1(builder, array_type_id, SpvDecorationArrayStride, sizeof(uint64_t));
+
+            pointer_struct_id = vkd3d_spirv_build_op_type_struct(builder, &array_type_id, 1);
+
+            vkd3d_spirv_build_op_member_decorate1(builder, pointer_struct_id, 0, SpvDecorationOffset, 0);
+            vkd3d_spirv_build_op_decorate(builder, pointer_struct_id, SpvDecorationBufferBlock, NULL, 0);
+            vkd3d_spirv_build_op_name(builder, pointer_struct_id, "uav_ctrs_t");
+
+            var_id = vkd3d_spirv_build_op_variable(builder, &builder->global_stream,
+                    vkd3d_spirv_get_op_type_pointer(builder, storage_class, pointer_struct_id),
+                    storage_class, 0);
```

Comment:
One missing piece here is that the SSBO variable should be declared with AliasedPointer decoration since the OpVariable holds physical pointers.

Reply:
See 2.18.2 in SPIR-V spec